typed-racket-lib/typed-racket/utils/hash-contract.rkt

```diff
@@ -1,8 +1,18 @@
 #lang racket/base
 
 (require racket/contract)
-(provide mutable-hash/c immutable-hash/c weak-hash/c)
+(provide tr-hash/c mutable-hash/c immutable-hash/c weak-hash/c)
 
-(define (mutable-hash/c k v) (and/c hash? (not/c hash-weak?) (hash/c k v #:immutable #f)))
-(define (immutable-hash/c k v) (hash/c k v #:immutable #t))
-(define (weak-hash/c k v) (and/c hash? hash-weak? (hash/c k v #:immutable #f)))
+(define (mutable-hash/c k v)
+  (and/c hash? (not/c hash-weak?) (tr-hash/c k v #:immutable #f)))
+
+(define (immutable-hash/c k v)
+  (tr-hash/c k v #:immutable #t))
+
+(define (weak-hash/c k v)
+  (and/c hash? hash-weak? (tr-hash/c k v #:immutable #f)))
+
+(define (tr-hash/c k v #:immutable [immutable 'dont-care])
+  (if (flat-contract? k)
+    (hash/c k v #:immutable immutable)
+    (and/c hash? hash-equal? (hash/c k v #:immutable immutable))))
```

Comment:
I think we need to do something to produce a good error message in this case if `hash-equal?` is `#f`, but I don't have a good suggestion.

Also, it would be even better if we could make this a contract that rejected the non-flat parts of `k` (probably we'd have to do that split in the static contract). But this fix, assuming it doesn't break things, is good for the release.

Reply:
Can we just let `hash/c` raise the error message for the release?

The only improvement I can think of is printing the key, e.g. "expected a hash-equal? because of key (vectorof foo?)"